hazelcast/src/main/java/com/hazelcast/internal/util/Timer.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.util;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Utility class to measure elapsed time in milliseconds. Backed by {@link System#nanoTime()}
+ * Example usage {@code
+ * long start = Timer.millis();
+ * ... // do some work
+ * long elapsedMs = Timer.elapsedSince(start);
+ * }
+ */
+public class Timer {
+    public static long millis() {
+        return TimeUnit.NANOSECONDS.toMillis(System.nanoTime());
+    }
+
+    public static long elapsedSince(long millis) {
```

Comment:
I believe this logic is flawed when there is an overflow.

Reproducer:

```
public class Main {

    public static void main(String[] args){
        withNanos();
        withMillis();
    }

    private static void withMillis() {
        long start = TimeUnit.NANOSECONDS.toMillis(Long.MAX_VALUE-10);
        long end = TimeUnit.NANOSECONDS.toMillis(Long.MAX_VALUE+1);
        long delta = end-start;
        System.out.println("delta "+delta+" ms");
    }

    private static void withNanos() {
        long start = Long.MAX_VALUE-10;
        long end = Long.MAX_VALUE+1;
        long delta = end-start;
        System.out.println("delta: "+delta+" ns");
    }
}
```

```
delta: 11 ns
delta -18446744073708 ms
```

Reply:
If your clock works on nanosecond level, and you deal with overflow correctly, it should be fine.

```
public class Timer {
    public static long nanos() {
        return System.nanoTime();
    }

    public static long nanosElapsedSince(long startNanos) {
        return System.nanoTime() - startNanos;
    }

    public static long millisElapsedSince(long startNanos) {
        return TimeUnit.NANOS.toMillis(System.nanoTime() - startNanos);
    }
}
```